controllers/reconcilers/grafana/http_route_reconciler.go

```diff
@@ -0,0 +1,300 @@
+package grafana
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/grafana/grafana-operator/v5/api/v1beta1"
+	"github.com/grafana/grafana-operator/v5/controllers/model"
+	"github.com/grafana/grafana-operator/v5/controllers/reconcilers"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	v2 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+// HTTPRouteReconciler is responsible for managing and reconciling
+// the Gateway API HTTPRoute resource associated with a Grafana instance.
+type HTTPRouteReconciler struct {
+	client client.Client
+}
+
+// NewHTTPRouteReconciler returns a new instance of HTTPRouteReconciler.
+func NewHTTPRouteReconciler(client client.Client) reconcilers.OperatorGrafanaReconciler {
+	return &HTTPRouteReconciler{
+		client: client,
+	}
+}
+
+// Reconcile runs the main reconciliation logic for Grafana HTTPRoute.
+// It ensures that the desired HTTPRoute exists and is configured correctly.
+func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, cr *v1beta1.Grafana, vars *v1beta1.OperatorReconcileVars, scheme *runtime.Scheme) (v1beta1.OperatorStageStatus, error) {
+	log := logf.FromContext(ctx).WithName("HTTPRouteReconciler")
+
+	log.Info("reconciling http route")
+
+	return r.reconcileHTTPRoute(ctx, cr, vars, scheme)
+}
+
+// reconcileHTTPRoute ensures the HTTPRoute object for Grafana matches the desired spec.
+// It creates or updates the HTTPRoute, merges configurations, and updates Grafana’s AdminURL if needed.
+func (r *HTTPRouteReconciler) reconcileHTTPRoute(ctx context.Context, cr *v1beta1.Grafana, _ *v1beta1.OperatorReconcileVars, scheme *runtime.Scheme) (v1beta1.OperatorStageStatus, error) {
+	if cr.Spec.HTTPRoute == nil {
+		return v1beta1.OperatorStageResultSuccess, nil
+	}
+
+	httpRoute := model.GetGrafanaHTTPRoute(cr, scheme)
+
+	_, err := controllerutil.CreateOrUpdate(ctx, r.client, httpRoute, func() error {
+		httpRoute.Spec = getHTTPRouteSpec(cr, scheme)
+
+		// Merge the CR-defined HTTPRoute into the generated object
+		err := v1beta1.Merge(cr.Spec.HTTPRoute, httpRoute)
+		if err != nil {
+			setInvalidMergeCondition(cr, "HTTPRoute", err)
+			return err
+		}
+
+		removeInvalidMergeCondition(cr, "HTTPRoute")
+
+		// Set Grafana as the controller owner
+		err = controllerutil.SetControllerReference(cr, httpRoute, scheme)
+		if err != nil {
+			return err
+		}
+
+		// Propagate labels from Grafana CR to the HTTPRoute
+		model.SetInheritedLabels(httpRoute, cr.Labels)
+
+		return nil
+	})
+	if err != nil {
+		return v1beta1.OperatorStageResultFailed, err
+	}
+
+	// Assign AdminURL if ingress is preferred
+	if cr.PreferIngress() {
```

Comment:
`Ingress` and `HTTPRoute` are separate resources, so I think there's a conflicting logic here.

Reply:
But if both `Ingress` and `HTTPRoute` are defined, there would be a conflicting back-and-forth change between `Ingress` and `HTTPRoute` reconcilers, right?